pkg/util/lsof/lsof_linux.go

```diff
@@ -0,0 +1,383 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package lsof
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// see the documentation for /proc for more details about files and their format
+// https://www.kernel.org/doc/html/latest/filesystems/proc.html
+
+// openFilesLister stores the state needed to list open files
+// this is helpful for tests to help with mocking
+type openFilesLister struct {
+	pid      int
+	procPath string
+
+	readlink func(string) (string, error)
+	stat     func(string) (os.FileInfo, error)
+	lstat    func(string) (os.FileInfo, error)
+
+	proc       procfsProc
+	socketInfo map[uint64]socketInfo
+}
+
+// procfsProc is an interface to allow mocking of procfs.Proc
+type procfsProc interface {
+	ProcMaps() ([]*procfs.ProcMap, error)
+	FileDescriptors() ([]uintptr, error)
+	Cwd() (string, error)
+}
+
+type socketInfo struct {
+	Description string
+	State       string
+	Protocol    string
+}
+
+func openFiles(_ context.Context, pid int) (Files, error) {
+	ofl := &openFilesLister{
+		pid: pid,
+
+		readlink: os.Readlink,
+		stat:     os.Stat,
+		lstat:    os.Lstat,
+	}
+
+	ofl.procPath = procPath()
+
+	fs, err := procfs.NewFS(ofl.procPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ofl.proc, err = fs.Proc(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	ofl.socketInfo = readSocketInfo(ofl.procPIDPath())
+
+	return ofl.openFiles()
+}
+
+func (ofl *openFilesLister) procPIDPath() string {
+	return fmt.Sprintf("%s/%d", ofl.procPath, ofl.pid)
+}
+
+func (ofl *openFilesLister) openFiles() (Files, error) {
+	var files Files
+
+	// open files, socket, pipe (everything with a file descriptor, from /proc/<pid>/fd)
+	openFDFiles, err := ofl.fdMetadata()
+	if err != nil {
+		log.Debugf("Failed to get open FDs for pid %d: %s", ofl.pid, err)
+	} else {
+		files = append(files, openFDFiles...)
+	}
+
+	// memory mapped files, code, regions (from /proc/<pid>/maps)
+	mmapFiles, err := ofl.mmapMetadata()
+	if err != nil {
+		log.Debugf("Failed to get memory maps for pid %d: %s", ofl.pid, err)
+	} else {
+		files = append(files, mmapFiles...)
+	}
+
+	return files, nil
+}
+
+func (ofl *openFilesLister) mmapMetadata() (Files, error) {
+	maps, err := ofl.proc.ProcMaps()
+	if err != nil {
+		return nil, err
+	}
+
+	cwd, _ := ofl.proc.Cwd()
+
+	var files Files
+	for i, m := range maps {
+		if i > 0 && m.Pathname == maps[i-1].Pathname {
```

Comment:
What would be a case of duplicate entry?

Reply:
There are a lot actually 😄 

The same file can be mapped to multiple regions in memory, possibly with different permissions...
You can check [pkg/util/lsof/testdata/mmapMetadata/1/maps](https://github.com/DataDog/datadog-agent/blob/pgimalac/lsof-flare/pkg/util/lsof/testdata/mmapMetadata/1/maps) for an example

I only keep the first entry as this is what `lsof` seem to do, it avoids polluting the output with dozens of duplicates